---
title: "Compare quantifications"
author: "Dave Tang"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(tximport)
knitr::opts_chunk$set(echo = TRUE)
```

## Setup

Use [tximport](https://bioconductor.org/packages/release/bioc/html/tximport.html) to load expression data.

```{r install_tximport, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("tximport")
```

## Data

For RSEM, `abundance` contains the TPM values. We will store the expression data into `rsem_gene` and `rsem_tran` for the gene and transcript quantifications, respectively.

```{r load_rsem, warning=FALSE, message=FALSE}
read_rsem_gene <- function(dir){
  rsem_gene_file <- list.files(
    path = dir,
    pattern = "\\.genes\\.results",
    recursive = TRUE,
    full.names = TRUE
  )
  my_sample <- sub(pattern = "(.*)_.*", replacement = "\\1", x = basename(rsem_gene_file))
  my_sample <- sub(pattern = ".genes.results", replacement = "", x = my_sample)
  rsem_gene <- tximport(
    rsem_gene_file,
    type = "rsem",
    txIn = FALSE,
    txOut = FALSE
  )
  colnames(rsem_gene$abundance) <- my_sample
  rsem_gene
}

read_rsem_tran <- function(dir){
  rsem_tran_file <- list.files(
    path = dir,
    pattern = "\\.isoforms\\.results",
    recursive = TRUE,
    full.names = TRUE
  )
  my_sample <- sub(pattern = "(.*)_.*", replacement = "\\1", x = basename(rsem_tran_file))
  my_sample <- sub(pattern = ".isoforms.results", replacement = "", x = my_sample)
  rsem_tran <- tximport(
    rsem_tran_file,
    type = "rsem",
    txIn = TRUE,
    txOut = TRUE
  )
  colnames(rsem_tran$abundance) <- my_sample
  rsem_tran
}

rsem_gene <- read_rsem_gene("../results/star_rsem")
rsem_tran <- read_rsem_tran("../results/star_rsem")

rsem_gene_nf <- read_rsem_gene("../results/nfcore_rnaseq/star_rsem/")
rsem_tran_nf <- read_rsem_tran("../results/nfcore_rnaseq/star_rsem/")

rsem_gene$abundance[1:6, 1:6]
rsem_tran$abundance[1:6, 1:6]
```

For StringTie, we can import the transcript quantification data stored in the `t_data.ctab` files. The format is as follows:

* t_id: numeric transcript id
* chr, strand, start, end: genomic location of the transcript
* t_name: Cufflinks-generated transcript id
* num_exons: number of exons comprising the transcript
* length: transcript length, including both exons and introns
* gene_id: gene the transcript belongs to
* gene_name: HUGO gene name for the transcript, if known
* cov: per-base coverage for the transcript (available for each sample)
* FPKM: Cufflinks-estimated FPKM for the transcript (available for each sample)

We can use the coverage data to obtain the original raw counts by calculating `cov * transcript length / read length`. The `tximport` function does not support importing `abundance.tsv.gz` files.

```{r load_stringtie, warning=FALSE, message=FALSE}
st_tran_file <- list.files(
  path = "../results/hisat_stringtie",
  pattern = "t_data\\.ctab$",
  recursive = TRUE,
  full.names = TRUE
)

# https://bioconductor.org/packages/devel/bioc/vignettes/tximport/inst/doc/tximport.html#StringTie
tmp <- read_tsv(st_tran_file[1], show_col_types = FALSE)
tx2gene <- tmp[, c("t_name", "gene_id")]

st_gene <- tximport(
  st_tran_file,
  tx2gene = tx2gene,
  type = "stringtie",
  txIn = TRUE,
  txOut = FALSE
)

st_tran <- tximport(
  st_tran_file,
  type = "stringtie",
  txIn = TRUE,
  txOut = TRUE
)

my_sample <- sub(pattern = ".*(ERR\\d+)_chrX.*", replacement = "\\1", x = st_tran_file)

colnames(st_gene$abundance) <- my_sample
colnames(st_tran$abundance) <- my_sample
st_tran$abundance[1:6, 1:6]
```

Calculate TPM values.

```{r st_tran_tpm}
calc_tpm <- function(obj){
  # divide read counts by the length of each gene in kilobases
  obj$tpm <- obj$counts / (obj$length / 1000)
  
  # sum all RPK values per sample and divide by 1,000,000
  tpm_scaling_factor <- apply(obj$tpm, 2, function(x) sum(x) / 1000000)
  
  # divide the RPK values by the scaling factor
  obj$tpm <- t(apply(obj$tpm, 1, function(x) x / tpm_scaling_factor))
  colnames(obj$tpm) <- my_sample
  obj$tpm
}

st_gene$tpm <- calc_tpm(st_gene)
st_tran$tpm <- calc_tpm(st_tran)
```

Load Kallisto results. (Note that if your directory contains HDF files `tximport` will try to read them regardless of specifying `abundance.tsv` files.)

```{r load_kallisto, warning=FALSE, message=FALSE}
kallisto_tran_file <- list.files(
  path = "../results/kallisto",
  pattern = "abundance\\.tsv",
  recursive = TRUE,
  full.names = TRUE
)

# caveat: using tx2gene created from StringTie results
kallisto_gene <- tximport(
  kallisto_tran_file,
  type = "kallisto",
  tx2gene = tx2gene,
  txIn = TRUE,
  txOut = FALSE
)

kallisto_tran <- tximport(
  kallisto_tran_file,
  type = "kallisto",
  txIn = TRUE,
  txOut = TRUE
)

my_sample <- sub(pattern = ".*(ERR\\d+)_chrX.*", replacement = "\\1", x = kallisto_tran_file)

colnames(kallisto_gene$abundance) <- my_sample
colnames(kallisto_tran$abundance) <- my_sample
kallisto_tran$abundance[1:6, 1:6]
```

## Comparisons

Correlate the gene expression values and plot results.

```{r correlate_genes}
my_gene <- row.names(rsem_gene$abundance)
my_sample <- colnames(rsem_gene$abundance)

purrr::map_df(my_sample, function(x){
  my_rsem <- rsem_gene$abundance[my_gene, x]
  my_st <- st_gene$tpm[my_gene, x]
  my_kallisto <- kallisto_gene$abundance[my_gene, x]
  
  rsem_vs_st <- cor(my_rsem, my_st, method = "spearman")
  rsem_vs_kallisto <- cor(my_rsem, my_kallisto, method = "spearman")
  st_vs_kallisto <- cor(my_st, my_kallisto, method = "spearman")
  
  list(
    RSEM_vs_StringTie = rsem_vs_st,
    RSEM_vs_Kallisto = rsem_vs_kallisto,
    StringTie_vs_Kallisto = st_vs_kallisto
  )
}) |>
  dplyr::mutate(sample = my_sample) |>
  dplyr::select(sample, everything()) -> gene_res

gene_res |>
  pivot_longer(-sample, names_to = "Comparison", values_to = "Spearman") |>
  ggplot(aes(Comparison, Spearman)) +
  geom_boxplot() +
  theme_minimal() +
  ggtitle("Spearman correlations between gene quantifications")
```

Correlate the transcript expression values and plot results.

```{r correlate_transcript}
my_gene <- row.names(rsem_tran$abundance)
my_sample <- colnames(rsem_tran$abundance)

purrr::map_df(my_sample, function(x){
  my_rsem <- rsem_tran$abundance[my_gene, x]
  my_st <- st_tran$tpm[my_gene, x]
  my_kallisto <- kallisto_tran$abundance[my_gene, x]
  
  rsem_vs_st <- cor(my_rsem, my_st, method = "spearman")
  rsem_vs_kallisto <- cor(my_rsem, my_kallisto, method = "spearman")
  st_vs_kallisto <- cor(my_st, my_kallisto, method = "spearman")
  
  list(
    RSEM_vs_StringTie = rsem_vs_st,
    RSEM_vs_Kallisto = rsem_vs_kallisto,
    StringTie_vs_Kallisto = st_vs_kallisto
  )
}) |>
  dplyr::mutate(sample = my_sample) |>
  dplyr::select(sample, everything()) -> transcript_res

transcript_res |>
  pivot_longer(-sample, names_to = "Comparison", values_to = "Spearman") |>
  ggplot(aes(Comparison, Spearman)) +
  geom_boxplot() +
  theme_minimal() +
  ggtitle("Spearman correlations between transcript quantifications")
```

Compare with `nf-core/rnaseq`.

```{r rsem_comp_gene}
my_gene <- row.names(rsem_gene_nf$abundance)
my_sample <- colnames(rsem_gene_nf$abundance)

purrr::map_dbl(my_sample, function(x){
  my_rsem <- rsem_gene$abundance[my_gene, x]
  my_rsem_nf <- rsem_gene_nf$abundance[my_gene, x]
  cor(my_rsem, my_rsem_nf, method = "spearman")
})
```

```{r rsem_comp_tran}
my_gene <- row.names(rsem_tran_nf$abundance)
my_sample <- colnames(rsem_tran_nf$abundance)

purrr::map_dbl(my_sample, function(x){
  my_rsem <- rsem_tran$abundance[my_gene, x]
  my_rsem_nf <- rsem_tran_nf$abundance[my_gene, x]
  cor(my_rsem, my_rsem_nf, method = "spearman")
})
```

```{r rsem_comp_plot}
my_gene <- row.names(rsem_gene_nf$abundance)
my_sample <- colnames(rsem_gene_nf$abundance)

purrr::map(my_sample, function(x){
  data.frame(
    simple = rsem_gene$abundance[my_gene, x],
    nfcore = rsem_gene_nf$abundance[my_gene, x]
  ) |>
    ggplot(aes(log1p(simple), log1p(nfcore))) +
    geom_point() +
    theme_minimal() +
    ggtitle(x) -> p
  print(p)
}) -> dev_null
```
